### ISSUE TITLE
Qubes Salt state module for downloading files to offline VMs

### ISSUE THREAD

**Issue by rootkovska** [QubesOS/qubes-issues#2162]
It would be useful in some scenarios to be able to state something like this (imagine this being part of e.g. `my-offline-vm.sls` state file):
```
get some tgz from the untrusted Web:
  qubes.get_file_via_dispvm:
    - source: http://somewhere.com/somefile.tgz
    - sha1: adc83b19e793491b1c6ea0fd8b46cd9f32e592fc
```

Salt would then spawn a DispVM (connected to the net, so this would not be a DispVM started from the VM in question, as in that case it would inherit network-not-having property), and then wget the requested files. Subsequently the file would be checked against the hash given, and if matches, copied to the target offline AppVM using qubes.Filecopy service.

This way we would be able to prepare custom offline AppVMs which require e.g. custom sources from the Web, such as e.g. an AppVM for running Electrum + Trezor.

**Comment by lkubb** [comment]
After some trial and error, I think I have a solution to this (already implemented + working, not sure if reasonably secure in its current state though).

1. A `qubes.FileRecv` service written in bash, which refuses to run on anything but a DVM. It has a single required argument, the hostname. scheme/path and file hash are read via stdin. It fetches the file using curl (similar to the template downloader), verifies the hash (if provided), then delegates to `qfile-agent`.
2. A modified salt.fileclient that uses this service + (unprivileged) `qfile-unpacker` to fetch URLs.
3. A custom `cp` module, which overrides the inbuilt one. It uses the modified fileclient if it detects it is running on a template qube or a qube with no network connection.
4. I have also written a very basic (Python) CLI client for this service. Might be useful, not sure.

Advantages:
* `file.managed` works on non-networked qubes more or less exactly the same way as on networked ones, no need for a separate state.
* Several other modules that fetch files from the network should also just work (most use the `cp` module for network transfers).
* The RPC policy can restrict access to specific domains only.

Drawbacks:
* Non-networked qubes can get network access if the policy allows it, so it's a service that can be risky.

I still need some time to evaluate it myself, then I'll see if I can come up with a PR, if desired.

**Comment by ben-grande** [comment]
@lkubb That is super cool!

> After some trial and error, I think I have a solution to this (already implemented + working, not sure if reasonably secure in its current state though).
> 
> 1. A `qubes.FileRecv` service written in bash, which refuses to run on anything but a DVM. It has a single required argument, the hostname. scheme/path and file hash are read via stdin. It fetches the file using curl (similar to the template downloader), verifies the hash (if provided), then delegates to `qfile-agent`.

I don't think Salt `file.managed` requires curl, I think it uses some python module to fetch files, as I have qubes without curl and Satl can fetch files on HTTP servers. Can you please drop `curl` and use the same python module Salt uses? This is important for users on minimal templates. Sorry for not providing more information on this, I didn't dig deeper and don't wan't to be demanding on your contribution.

The hash needs to be verified on the qube that wants the file, not the one that downloads the file, this way, the qube that wants the file doesn't need to trust the disposable that downloaded the qube. Per your explanation, it seems that checksum happens on the disposable qube, as you placed `qfile-agent` at last.

> 2. A modified salt.fileclient that uses this service + (unprivileged) `qfile-unpacker` to fetch URLs.
> 3. A custom `cp` module, which overrides the inbuilt one. It uses the modified fileclient if it detects it is running on a template qube or a qube with no network connection.
> 4. I have also written a very basic (Python) CLI client for this service. Might be useful, not sure.

It seems useful when not calling via Salt in Dom0, but inside the qube via terminal.

> Advantages:
> 
> * `file.managed` works on non-networked qubes more or less exactly the same way as on networked ones, no need for a separate state.
> * Several other modules that fetch files from the network should also just work (most use the `cp` module for network transfers).
> * The RPC policy can restrict access to specific domains only.

I believe this is incorrect use of the RPC policy. The domain can already be restricted via policy being it the source or destination of a call. The argument can either be empty in this case or serve for another use case.
 
> Drawbacks:
> 
> * Non-networked qubes can get (indirect) network access if the policy allows it, so it's a service that can be risky.
> 
> I still need some time to evaluate it myself, then I'll see if I can come up with a PR, if desired.

Desired!

**Comment by ben-grande** [comment]
> The hash needs to be verified on the qube that wants the file, not the one that downloads the file, this way, the qube that wants the file doesn't need to trust the disposable that downloaded the qube. Per your explanation, it seems that checksum happens on the disposable qube, as you placed qfile-agent at last.

The hash can be checked on the disposable to see if file was partially downloaded or is corrupt, but the checksum can't be trusted on the qube requesting the file, it must be done again. The checksum on different qubes serve different purposes.

**Comment by lkubb** [comment]
Thanks for your reply @ben-grande  :)

> I don't think Salt `file.managed` requires curl, I think it uses some python module to fetch files, as I have qubes without curl and Satl can fetch files on HTTP servers. Can you please drop `curl` and use the same python module Salt uses? This is important for users on minimal templates.

`curl` is required on the spawned DVM that executes the service, not on the managed Qube - this works on minimal templates (I use them myself). If it used Salt's mechanism, it would require Python + `tornado`/`urllib2`/`requests`. A more detailed explanation:

1. `file.managed` calls `cp.cache_file` (the overridden one in the custom module)
2. `cp.cache_file` delegates to the patched fileclient
3. DVM is spawned for executing the service
4. DVM runs `curl`, checks hash, runs `qfile-agent` if everything is OK
5. managed VM runs `qfile-unpacker` (as the default user, not root)
6. if the transfer succeeds, it checks the hash again
7. if OK, the file is moved to the final Salt cache dir

> The hash needs to be verified on the qube that wants the file, not the one that downloads the file, this way, the qube that wants the file doesn't need to trust the disposable that downloaded the qube.

Yes, my description is incomplete in this regard. It is checked both on the DVM and the managed Qube.

> I believe this is incorrect use of the RPC policy. The domain can already be restricted via policy being it the source or destination of a call. The argument can either be empty in this case or serve for another use case.

By domain, I mean the host name part of the URL, not another Qube. Sorry for the confusion!

> Desired!

Awesome!

**Comment by ben-grande** [comment]
> > I don't think Salt `file.managed` requires curl, I think it uses some python module to fetch files, as I have qubes without curl and Satl can fetch files on HTTP servers. Can you please drop `curl` and use the same python module Salt uses? This is important for users on minimal templates.
> 
> `curl` is required on the spawned DVM that executes the service, not on the managed Qube - this works on minimal templates (I use them myself). If it used Salt's mechanism, it would require Python + `tornado`/`urllib2`/`requests`.

Great!

> A more detailed explanation:
>
> 1. `file.managed` calls `cp.cache_file` (the overridden one in the custom module)
> 2. `cp.cache_file` delegates to the patched fileclient
> 3. DVM is spawned for executing the service
> 4. DVM runs `curl`, checks hash, runs `qfile-agent` if everything is OK
> 5. managed VM runs `qfile-unpacker` (as the default user, not root)
> 6. if the transfer succeeds, the patched fileclient on the managed VM checks the hash again
> 7. if OK, it moves the file to the final Salt cache dir on the managed VM

The flow is clearer now, thanks.

---

About the other points, it was just my interpretation, with your clarifications. everything seems right.

**Comment by ben-grande** [comment]
Hi. Me again.

Salt 3007.0, new options to  [archive.extracted](https://docs.saltproject.io/en/latest/ref/states/all/salt.states.archive.html#salt.states.archive.extracted) and [gpg.verify](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.gpg.html#salt.modules.gpg.verify) modules have been added regarding PGP signature verification, most notably:

- You can trust only certain keys to be able to verify a file, the verification can occur by any key or enforce that all keys needs to be used to verify a file (`signed_by_any`, `signed_by_all`.
- The archive.extracted can now also choose the `gnupghome` and `keyring`.
- The signature verification can occur on the checksum file `source_hash_sig`

Unfortunately it appears to be hardcoded to the `gpg` binary, so  using [Sequoia Chameleon](https://sequoia-pgp.org/blog/2022/12/19/202212-chameleon-0.1/) for a more modern approach and codebase does not seem possible.

[Note on hardcoded binary](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.gpg.html#module-salt.modules.gpg):
*The python-gnupg library and gpg binary are required to be installed. Be aware that the alternate gnupg and pretty-bad-protocol libraries are not supported.*

Although not necessary for the first version, I'd leave this as an enhancement for the future.

**Comment by lkubb** [comment]
@ben-grande 

Yup, all these features were my work as well. ;) They also work with this enhancement btw.

Interesting, haven't heard of Sequoia so far. As mentioned, the `gpg` modules use `python-gnupg` internally. There is some info regarding the use of Sequoia: https://github.com/vsajip/python-gnupg/issues/217

Forcing the module to load should be somewhat trivial, but the issue mentions that not all `python-gnupg` tests were passing at that time. We could investigate what is missing, but it might take me a while since I'm focusing hard on an improved Qubes-Salt integration atm.
